File: 2695-find-score-of-an-array-after-marking-all-elements/2695-find-score-of-an-array-after-marking-all-elements.py
from collections import deque
import heapq

class Tuple:
    def __init__(self, val, idx):
        self.val = val
        self.idx = idx
    
    def __lt__(self, elem):
        if(self.val < elem.val): return True
        elif(self.val == elem.val): return self.idx < elem.idx
        else: return False

class Solution:
    def findScore(self, nums: List[int]) -> int:
        n = len(nums)
        elems = []

        for idx, val in enumerate(nums):
            elems.append(Tuple(val, idx))

        heapq.heapify(elems)
        ans = 0
        while len(elems) > 0:
            elem = heapq.heappop(elems)
            print(f"{elem.val} and {elem.idx}")
            idx = elem.idx
            val = nums[idx]

            if(val == 0): continue

            ans = ans + val

            leftIdx = idx - 1
            rightIdx = idx + 1

            if(leftIdx >= 0): nums[leftIdx] = 0
            if(rightIdx < n): nums[rightIdx] = 0

        return ans

File: 2695-find-score-of-an-array-after-marking-all-elements/2695-find-score-of-an-array-after-marking-all-elements.cpp
class Comp{
    public:
    bool operator()(pair<int, int> p1, pair<int, int> p2){
        if(p1.first > p2.first) return true;
        else if(p1.first == p2.first) return p1.second > p2.second;
        return false;
    }
};

class Solution {
public:

    long long findScore(vector<int>& nums) {
        int n = nums.size();
        long long ans = 0;
        priority_queue<pair<int, int>, vector<pair<int, int>>, Comp> minPQ;
        
        for(int i = 0; i < n; i++){
            minPQ.push(make_pair(nums[i], i));
        }

        while(!minPQ.empty()){
            pair<int,int> topItem = minPQ.top(); minPQ.pop();
            
            
            int idx = topItem.second;
            int val = nums[idx];

            cout << idx << " having value " << val << endl;

            if(val == 0) continue;
            ans = ans + (long long)(val);
            
            int leftIdx = idx-1;
            int rightIdx = idx+1;

            if(leftIdx >= 0) nums[leftIdx] = 0;
            if(rightIdx < n) nums[rightIdx] = 0;
        }

        return ans;
    }
};